param/parameterized.py

```diff
@@ -2826,13 +2831,135 @@ def log(self_, level, msg, *args, **kw):
+                # change.
+            elif hasattr(g,'state_push') and isinstance(g,Parameterized):
+                g.state_push()
+
+    def _state_pop(self_):
+        """
+        Restore the most recently saved state.
+
+        See state_push() for more details.
+        """
+        self = self_.self_or_cls
+        if not isinstance(self, Parameterized):
+            raise NotImplementedError('_state_pop is not implemented at the class level')
+        for pname, p in self.param.objects('existing').items():
+            g = self.param.get_value_generator(pname)
+            if hasattr(g,'_Dynamic_last'):
+                g._Dynamic_last = g._saved_Dynamic_last.pop()
+                g._Dynamic_time = g._saved_Dynamic_time.pop()
+            elif hasattr(g,'state_pop') and isinstance(g,Parameterized):
+                g.state_pop()
 
+    @_recursive_repr()
     def pprint(self_, imports=None, prefix=" ", unknown_value='<?>',
                qualify=False, separator=""):
-        """See Parameterized.pprint"""
-        self = self_.self
-        return self._pprint(imports, prefix, unknown_value, qualify, separator)
+        """
+        (Experimental) Pretty printed representation that may be
+        evaluated with eval. See pprint() function for more details.
+        """
+        self = self_.self_or_cls
+        if not isinstance(self, Parameterized):
+            raise NotImplementedError('_pprint is not implemented at the class level')
+        if imports is None:
+            imports = [] # would have been simpler to use a set from the start
+        imports[:] = list(set(imports))
 
+        # Generate import statement
+        mod = self.__module__
+        bits = mod.split('.')
+        imports.append("import %s"%mod)
+        imports.append("import %s"%bits[0])
+
+        changed_params = self.param.values(onlychanged=script_repr_suppress_defaults)
+        values = self.param.values()
+        spec = getfullargspec(self.__init__)
+        args = spec.args[1:] if spec.args[0] == 'self' else spec.args
+
+        if spec.defaults is not None:
+            posargs = spec.args[:-len(spec.defaults)]
+            kwargs = dict(zip(spec.args[-len(spec.defaults):], spec.defaults))
+        else:
+            posargs, kwargs = args, []
+
+        parameters = self.param.objects('existing')
+        ordering = sorted(
+            sorted(changed_params), # alphanumeric tie-breaker
+            key=lambda k: (- float('inf')  # No precedence is lowest possible precendence
+                           if parameters[k].precedence is None else
+                           parameters[k].precedence))
+
+        arglist, keywords, processed = [], [], []
+        for k in args + ordering:
+            if k in processed: continue
+
+            # Suppresses automatically generated names.
+            if k == 'name' and (values[k] is not None
+                                and re.match('^'+self.__class__.__name__+'[0-9]+$', values[k])):
```

Comment:
I think Regex is overkill here. But I can't seem to get a better speedup with other methods for a simple example. Maybe it is caching...

It could be improved by using `re.compile` outside the for loop. Maybe also limit the numbers to 5.